HPCC-20023 ESDL binding loading issues

mayx opened the pull request:
- Make the constructed esp configuration for dynamic loading more complete
- Check whether template binding is configured at the time of adding the
  binding
- Use numeric value to search by port in property tree

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

mayx commented:
@rpastrana please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 71320c12324fb14a9944aee59a756aa46b0d51cd
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |789 |789 |0 |
|test (thor) |720 |720 |0 |
|test (roxie) |860 |860 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   27 sec (00:00:27) |  171 sec (00:02:51) |   61 sec (00:01:01) |    9 sec (00:00:09) |   30 sec (00:00:30) | 1284 sec (00:21:24) |   18 sec (00:00:18) | 1600 sec (00:26:40) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: baacae170c831d156bc6c22270f8ff3188474301
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |789 |789 |0 |
|test (thor) |720 |720 |0 |
|test (roxie) |860 |860 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   12 sec (00:00:12) |  170 sec (00:02:50) |   62 sec (00:01:02) |    9 sec (00:00:09) |   30 sec (00:00:30) | 1274 sec (00:21:14) |   19 sec (00:00:19) | 1576 sec (00:26:16) |

rpastrana reviewed (commented):
@mayx a few minor comments

mayx commented on esp/services/esdl_svc_engine/esdl_monitor.cpp:
```diff
@@ -454,23 +454,17 @@ class CEsdlMonitor : implements IEsdlMonitor, public CInterface, implements IEsd
+            Owned<IPropertyTree> proctemplate = createPTreeFromIPT(procpt);
+            proctemplate->removeProp("EspBinding");
+            proctemplate->removeProp("EspService");
+            m_envptTemplate->addPropTree("Software/EspProcess", proctemplate.getClear());
```
Yes it's ESPProcess without EspBindings and EspServices. I tested removing doesn't cause ambiguous path errors.

rpastrana replied:
I'm surprised there's no issue when multiple EspBinding (or EspService) branches exists...
does removeProp remove all instances or does it only remove one?

mayx replied:
removeProp removes all instances of EspBinding and EspService.

mayx commented on esp/services/esdl_svc_engine/esdl_monitor.cpp:
```diff
@@ -498,12 +492,12 @@ class CEsdlMonitor : implements IEsdlMonitor, public CInterface, implements IEsd
             }
             //Otherwise check if there's binding configured on the same port
-            xpath.clear().appendf("EspBinding[@type='EsdlBinding'][@port='%s'][1]", port);
+            xpath.clear().appendf("EspBinding[@type='EsdlBinding'][@port=%s][1]", port);
```
fixed.

mayx replied:
As for the port, yes since port is numeric, comparing numerically will help with cases when someone manually edits the environment.xml file and uses "00" instead "0" for example.

mayx commented on esp/services/esdl_svc_engine/esdl_monitor.cpp:
```diff
@@ -498,12 +492,12 @@ class CEsdlMonitor : implements IEsdlMonitor, public CInterface, implements IEsd
             {
                 //Otherwise check if there's binding configured with port 0
-                xpath.clear().appendf("EspBinding[@type='EsdlBinding'][@port='0']");
+                xpath.clear().appendf("EspBinding[@type='EsdlBinding'][@port=0]");
```
fixed.

mayx commented on initfiles/etc/DIR_NAME/environment.xml.in:
```diff
@@ -489,6 +489,26 @@
+               protocol="http"
+               resourcesBasedn="ou=DynamicESDL,ou=EspServices,ou=ecl"
+               service="DESDLServiceTemplate"
+               type=""
```
It seems all other bindings use empty string, so let's keep it this way to be consistent...

mayx commented:
@rpastrana change made per comments, please review again...

mayx committed:
HPCC-20023 ESDL binding loading issues

- Make the constructed esp configuration for dynamic loading more complete
- Check whether template binding is configured at the time of adding the
  binding
- Use numeric value to search by port in property tree
- Add a default DESDL binding and service template to environment.xml

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 099df2a5dbfb89503eed569c0dd64e360fc3a318
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |789 |789 |0 |
|test (thor) |720 |720 |0 |
|test (roxie) |860 |860 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   16 sec (00:00:16) |  170 sec (00:02:50) |   62 sec (00:01:02) |    7 sec (00:00:07) |   27 sec (00:00:27) | 1291 sec (00:21:31) |   18 sec (00:00:18) | 1591 sec (00:26:31) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 63daf1981870c7fdca6804f95e4359af46a32f95
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |789 |789 |0 |
|test (thor) |720 |720 |0 |
|test (roxie) |860 |860 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   18 sec (00:00:18) |  173 sec (00:02:53) |   61 sec (00:01:01) |    6 sec (00:00:06) |   28 sec (00:00:28) | 1250 sec (00:20:50) |   20 sec (00:00:20) | 1556 sec (00:25:56) |

rpastrana reviewed (commented):
@mayx looks fine. I'm just concerned about the removeprop, if you can confirm it behaves as intended when given an ambiguous path

mayx commented:
@rpastrana I double checked that removeProp works as expected, please approve the PR.

rpastrana reviewed (approved):
@mayx
